lib/spack/docs/tutorial_modules.rst

```diff
@@ -776,113 +1103,169 @@ After modifications the configuration file will be:
 
-^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
-Extend the hierarchy to other virtual providers
-^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Add LAPACK to the hierarchy
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The hierarchy just shown is already a great improvement over non-hierarchical layouts,
+but it still has an asymmetry: ``LAPACK`` providers cover the same semantic role
+as ``MPI`` providers, but yet they are not part of the hierarchy.
+
+To be more practical, this means that despite we have gained an improved consistency in
+our environment when it comes to ``MPI``, we still have the same problems as we had before
+for ``LAPACK`` implementations:
 
-.. warning::
-  This is an experimental feature
-    Having a hierarchy deeper than ``Core``/``Compiler``/``MPI`` is an experimental
-    feature, still not fully supported by ``module spider``,
-    see `here <https://github.com/TACC/Lmod/issues/114>`_. Furthermore its use
-    with hierarchies more complex than ``Core``/``Compiler``/``MPI``/``LAPACK``
-    has not been thoroughly tested in production environments.
+.. code-block:: console
+
+  root@module-file-tutorial:/# module list
+
+  Currently Loaded Modules:
+    1) gcc/7.2.0   2) openblas/0.2.20   3) openmpi/1.10.2   4) netlib-scalapack/2.0.2-openblas
 
-Spack permits you to generate Lua hierarchical module files where users
-can add an arbitrary list of virtual providers to the triplet
-``Core``/``Compiler``/``MPI``. A configuration file like:
+  root@module-file-tutorial:/# module load netlib-scalapack/2.0.2-netlib
+  Autoloading netlib-lapack/3.6.1
+
+  The following have been reloaded with a version change:
+    1) netlib-scalapack/2.0.2-openblas => netlib-scalapack/2.0.2-netlib
+
+  root@module-file-tutorial:/# module list
+
+  Currently Loaded Modules:
+    1) gcc/7.2.0   2) openblas/0.2.20   3) openmpi/1.10.2   4) netlib-lapack/3.6.1   5) netlib-scalapack/2.0.2-netlib
+
+Now: having hierarchies that are deeper than ``Core``/``Compiler``/``MPI`` is
+probably still considered "unusual" or "impractical" at many sites, mainly because
+module files are written manually and keeping track of the combinations
+among multiple providers becomes quickly quite involved.
+
+For instance, having both ``MPI`` and ``LAPACK`` in the hierarchy
+means we must classify software into one of four categories:
+
+  #. Software that doesn't depend on ``MPI`` or ``LAPACK``
+  #. Software that depends only on ``MPI``
+  #. Software that depends only on ``LAPACK``
+  #. Software that depends on both
+
+to decide when to show it to the user. The situation becomes more involved as the number of virtual
+dependencies in the hierarchy increases.
+
+Anyhow in Spack we can take advantage of the internal DAG representation that any of
+the installed software has, and solve this combinatorial problem in a clean and automated way.
+In some sense deeper hierarchies are feasible due to Spack's ability to manage combinatorial complexity!
```

Comment:
Spack's ability to manage this combinatorial complexity makes deeper hiearchies feasible.